rust-runtime/aws-smithy-http/src/byte_stream/bytestream_util.rs

```diff
@@ -150,32 +171,54 @@ impl FsBuilder {
         let buffer_size = self.buffer_size;
 
         if let Some(path) = self.path {
-            let path_buf = path.to_path_buf();
-            let file_size = self.file_size.unwrap_or(
-                tokio::fs::metadata(path)
+            let offset = self.offset.unwrap_or(DEFAULT_OFFSET);
+            let length = self.length.unwrap_or({
+                let file_length = tokio::fs::metadata(&path)
                     .await
                     .map_err(|err| Error(err.into()))?
-                    .len(),
-            );
+                    .len();
 
+                // Length should never be less than zero. If the provided offset is greater than
+                // the length of the file, the number of bytes to read will be set to 0.
+                file_length.saturating_sub(offset)
+            });
+
+            let path_buf = path.to_path_buf();
             let body_loader = move || {
+                // If an offset was provided, seeking will be handled in `PathBody::poll_data` each
+                // time the file is loaded.
                 SdkBody::from_dyn(http_body::combinators::BoxBody::new(PathBody::from_path(
                     path_buf.clone(),
-                    file_size,
+                    length,
                     buffer_size,
+                    self.offset,
                 )))
             };
             Ok(ByteStream::new(SdkBody::retryable(body_loader)))
-        } else if let Some(file) = self.file {
-            let file_size = self.file_size.unwrap_or(
-                file.metadata()
+        } else if let Some(mut file) = self.file {
+            let offset = self.offset.unwrap_or(DEFAULT_OFFSET);
+            let length = self.length.unwrap_or({
```

Comment:
same here

Reply:
I guess _technically_ this works because the future is never polled?

Reply:
I switched this one to a match statement too